# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/CanonicalBrowsingContext.h
# Commit: 6fd57b83bb52
# Full Hash: 6fd57b83bb52e9683bade8cbfb5f8ba6575ac6af
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-02-07 03:55:51
# Regressor Bug: 1582832
# File Overlap Count: 3
# Description:
#   Bug 1582832: Part 1 - Make FrameLoader owner rather than DocShell responsible for discarding a BC. r=nika
#   
#   There are all sorts of lifecycle issues which arise from making DocShell
#   responsible for discarding BrowsingContexts. In this particular bug, we tend
#   to run into them in cases where we create a BrowsingContext for a FrameLoader,
# ==============================================================================

diff -r 28a9e4b075af -r 6fd57b83bb52 docshell/base/CanonicalBrowsingContext.h
--- a/docshell/base/CanonicalBrowsingContext.h	Thu Feb 06 18:35:10 2020 +0000
+++ b/docshell/base/CanonicalBrowsingContext.h	Thu Feb 06 19:07:56 2020 +0000
@@ -41,7 +41,11 @@
   bool IsOwnedByProcess(uint64_t aProcessId) const {
     return mProcessId == aProcessId;
   }
+  bool IsEmbeddedInProcess(uint64_t aProcessId) const {
+    return mEmbedderProcessId == aProcessId;
+  }
   uint64_t OwnerProcessId() const { return mProcessId; }
+  uint64_t EmbedderProcessId() const { return mEmbedderProcessId; }
   ContentParent* GetContentParent() const;
 
   void GetCurrentRemoteType(nsAString& aRemoteType, ErrorResult& aRv) const;
@@ -113,8 +117,10 @@
   using Type = BrowsingContext::Type;
   CanonicalBrowsingContext(BrowsingContext* aParent,
                            BrowsingContextGroup* aGroup,
-                           uint64_t aBrowsingContextId, uint64_t aProcessId,
-                           Type aType, FieldTuple&& aFields);
+                           uint64_t aBrowsingContextId,
+                           uint64_t aOwnerProcessId,
+                           uint64_t aEmbedderProcessId, Type aType,
+                           FieldTuple&& aFields);
 
  private:
   friend class BrowsingContext;
@@ -149,6 +155,9 @@
   // Indicates which process owns the docshell.
   uint64_t mProcessId;
 
+  // Indicates which process owns the embedder element.
+  uint64_t mEmbedderProcessId;
+
   // The ID of the former owner process during an ownership change, which may
   // have in-flight messages that assume it is still the owner.
   uint64_t mInFlightProcessId = 0;